File: python_sdk/src/reality_capture/service/job.py
import urllib.parse
from pydantic import BaseModel, Field, ValidationInfo, field_validator
from enum import Enum
from typing import Union, Optional, Any

from reality_capture.common.job import JobType, JobState, BaseProgress, BaseExecution
from reality_capture.specifications.calibration import CalibrationSpecifications, CalibrationSpecificationsCreate
from reality_capture.specifications.change_detection import (ChangeDetectionSpecifications,
                                                             ChangeDetectionSpecificationsCreate)
from reality_capture.specifications.constraints import (ConstraintsSpecificationsCreate,
                                                        ConstraintsSpecifications)
from reality_capture.specifications.fill_image_properties import (FillImagePropertiesSpecificationsCreate,
                                                                  FillImagePropertiesSpecifications)
from reality_capture.specifications.import_point_cloud import ImportPCSpecifications, ImportPCSpecificationsCreate
from reality_capture.specifications.objects2d import Objects2DSpecifications, Objects2DSpecificationsCreate
from reality_capture.specifications.production import ProductionSpecifications, ProductionSpecificationsCreate
from reality_capture.specifications.reconstruction import (ReconstructionSpecifications,
                                                           ReconstructionSpecificationsCreate)
from reality_capture.specifications.segmentation2d import (Segmentation2DSpecifications,
                                                           Segmentation2DSpecificationsCreate)
from reality_capture.specifications.segmentation3d import (Segmentation3DSpecifications,
                                                           Segmentation3DSpecificationsCreate)
from reality_capture.specifications.segmentation_orthophoto import (SegmentationOrthophotoSpecifications,
                                                                    SegmentationOrthophotoSpecificationsCreate)
from reality_capture.specifications.tiling import TilingSpecifications, TilingSpecificationsCreate
from reality_capture.specifications.touchup import (TouchUpImportSpecifications, TouchUpImportSpecificationsCreate,
                                                    TouchUpExportSpecifications, TouchUpExportSpecificationsCreate)
from reality_capture.specifications.water_constraints import (WaterConstraintsSpecifications,
                                                              WaterConstraintsSpecificationsCreate)
"""from reality_capture.specifications.point_cloud_conversion import (PointCloudConversionSpecificationsCreate,
                                                                   PointCloudConversionSpecifications)"""

from reality_capture.specifications.gaussian_splats import (GaussianSplatsSpecificationsCreate,
                                                            GaussianSplatsSpecifications)
from reality_capture.specifications.eval_o2d import (EvalO2DSpecificationsCreate, EvalO2DSpecifications)
from reality_capture.specifications.eval_o3d import (EvalO3DSpecificationsCreate, EvalO3DSpecifications)
from reality_capture.specifications.eval_s2d import (EvalS2DSpecificationsCreate, EvalS2DSpecifications)
from reality_capture.specifications.eval_s3d import (EvalS3DSpecificationsCreate, EvalS3DSpecifications)
from reality_capture.specifications.eval_sortho import (EvalSOrthoSpecificationsCreate, EvalSOrthoSpecifications)

from reality_capture.specifications.clearance import (ClearanceSpecificationsCreate, ClearanceSpecifications)

from reality_capture.service.reality_data import URL


class Service(Enum):
    MODELING = "Modeling"
    ANALYSIS = "Analysis"
    # CONVERSION = "Conversion"


def _get_appropriate_service(jt: JobType):
    if jt in [JobType.FILL_IMAGE_PROPERTIES, JobType.IMPORT_POINT_CLOUD, JobType.CALIBRATION, JobType.TILING,
              JobType.PRODUCTION, JobType.RECONSTRUCTION, JobType.CONSTRAINTS, JobType.TOUCH_UP_EXPORT,
              JobType.TOUCH_UP_IMPORT, JobType.WATER_CONSTRAINTS, JobType.GAUSSIAN_SPLATS]:
        return Service.MODELING
    if jt in [JobType.OBJECTS_2D, JobType.SEGMENTATION_2D, JobType.SEGMENTATION_3D, JobType.SEGMENTATION_ORTHOPHOTO,
              JobType.CHANGE_DETECTION, JobType.EVAL_O2D, JobType.EVAL_O3D, JobType.EVAL_S2D,
              JobType.EVAL_S3D, JobType.EVAL_SORTHO, JobType.CLEARANCE_CALCULATION]:
        return Service.ANALYSIS
    # return Service.CONVERSION
    raise NotImplementedError("Other services not yet implemented")


class JobCreate(BaseModel):
    name: Optional[str] = Field(None, description="Displayable job name.", min_length=3)
    type: JobType = Field(description="Type of job.")
    # TODO : PointCloudConversionSpecificationsCreate,
    specifications: Union[CalibrationSpecificationsCreate, ChangeDetectionSpecificationsCreate,
                          ConstraintsSpecificationsCreate,
                          EvalO2DSpecificationsCreate, EvalO3DSpecificationsCreate,
                          EvalS2DSpecificationsCreate, EvalS3DSpecificationsCreate,
                          EvalSOrthoSpecificationsCreate, FillImagePropertiesSpecificationsCreate,
                          GaussianSplatsSpecificationsCreate, ImportPCSpecificationsCreate,
                          Objects2DSpecificationsCreate, ProductionSpecificationsCreate,
                          ReconstructionSpecificationsCreate, Segmentation2DSpecificationsCreate,
                          Segmentation3DSpecificationsCreate, SegmentationOrthophotoSpecificationsCreate,
                          TilingSpecificationsCreate, TouchUpExportSpecificationsCreate,
                          TouchUpImportSpecificationsCreate, WaterConstraintsSpecificationsCreate,
                          ClearanceSpecificationsCreate] = (
        Field(description="Specifications aligned with the job type."))
    itwin_id: str = Field(description="iTwin ID, used by the service for finding "
                                      "input reality data and uploading output data.",
                          alias="iTwinId")

    def get_appropriate_service(self) -> Service:
        """
        Return the appropriate service for such a job.

        :return: Service enum.
        """
        return _get_appropriate_service(self.type)


class Execution(BaseExecution):
    processing_units: Optional[float] = Field(None, description="Processing units consumed by the job.",
                                              alias="processingUnits")


class Job(BaseModel):
    id: str = Field(description="Job unique identifier.")
    name: Optional[str] = Field(None, description="Displayable job name.", min_length=3)
    type: JobType = Field(description="Type of job.")
    itwin_id: str = Field(description="iTwin ID, used by the service for finding "
                                      "input reality data and uploading output data.",
                          alias="iTwinId")
    state: JobState = Field(description="State of the job.")
    execution_info: Execution = Field(description="Known execution information for the job.", alias="executionInfo")
    user_id: str = Field(description="Identifier of the user that created the job.", alias="userId")
    # TODO : add PointCloudConversionSpecifications
    specifications: Union[CalibrationSpecifications, ChangeDetectionSpecifications,
                          ConstraintsSpecifications,
                          EvalO2DSpecifications, EvalO3DSpecifications,
                          EvalS2DSpecifications, EvalS3DSpecifications,
                          EvalSOrthoSpecifications, FillImagePropertiesSpecifications,
                          GaussianSplatsSpecifications, ImportPCSpecifications,
                          Objects2DSpecifications, ProductionSpecifications,
                          ReconstructionSpecifications, Segmentation2DSpecifications,
                          Segmentation3DSpecifications, SegmentationOrthophotoSpecifications,
                          TilingSpecifications, TouchUpExportSpecifications,
                          TouchUpImportSpecifications, WaterConstraintsSpecifications,
                          ClearanceSpecifications] = (
        Field(description="Specifications aligned with the job type."))

    @field_validator("specifications", mode="plain")
    @classmethod
    def set_specification_validation_model(cls, raw_dict: dict[str, Any], validation_info: ValidationInfo):
        job_type = validation_info.data['type']

        specifications = None

        if job_type == JobType.CALIBRATION:
            specifications = CalibrationSpecifications(**raw_dict)
        elif job_type == JobType.CHANGE_DETECTION:
            specifications = ChangeDetectionSpecifications(**raw_dict)
        elif job_type == JobType.CONSTRAINTS:
            specifications = ConstraintsSpecifications(**raw_dict)
        elif job_type == JobType.EVAL_O2D:
            specifications = EvalO2DSpecifications(**raw_dict)
        elif job_type == JobType.EVAL_O3D:
            specifications = EvalO3DSpecifications(**raw_dict)
        elif job_type == JobType.EVAL_S2D:
            specifications = EvalS2DSpecifications(**raw_dict)
        elif job_type == JobType.EVAL_S3D:
            specifications = EvalS3DSpecifications(**raw_dict)
        elif job_type == JobType.EVAL_SORTHO:
            specifications = EvalSOrthoSpecifications(**raw_dict)
        elif job_type == JobType.FILL_IMAGE_PROPERTIES:
            specifications = FillImagePropertiesSpecifications(**raw_dict)
        elif job_type == JobType.GAUSSIAN_SPLATS:
            specifications = GaussianSplatsSpecifications(**raw_dict)
        elif job_type == JobType.IMPORT_POINT_CLOUD:
            specifications = ImportPCSpecifications(**raw_dict)
        elif job_type == JobType.OBJECTS_2D:
            specifications = Objects2DSpecifications(**raw_dict)
        elif job_type == JobType.PRODUCTION:
            specifications = ProductionSpecifications(**raw_dict)
        elif job_type == JobType.RECONSTRUCTION:
            specifications = ReconstructionSpecifications(**raw_dict)
        elif job_type == JobType.SEGMENTATION_2D:
            specifications = Segmentation2DSpecifications(**raw_dict)
        elif job_type == JobType.SEGMENTATION_3D:
            specifications = Segmentation3DSpecifications(**raw_dict)
        elif job_type == JobType.SEGMENTATION_ORTHOPHOTO:
            specifications = SegmentationOrthophotoSpecifications(**raw_dict)
        elif job_type == JobType.TILING:
            specifications = TilingSpecifications(**raw_dict)
        elif job_type == JobType.TOUCH_UP_EXPORT:
            specifications = TouchUpExportSpecifications(**raw_dict)
        elif job_type == JobType.TOUCH_UP_IMPORT:
            specifications = TouchUpImportSpecifications(**raw_dict)
        elif job_type == JobType.WATER_CONSTRAINTS:
            specifications = WaterConstraintsSpecifications(**raw_dict)
        elif job_type == JobType.CLEARANCE_CALCULATION:
            specifications = ClearanceSpecifications(**raw_dict)
        else:
            raise ValueError(f"Unsupported job type: {job_type}")

        return specifications

    def get_appropriate_service(self) -> Service:
        """
        Return the appropriate service for such a job.

        :return: Service enum.
        """
        return _get_appropriate_service(self.type)


class JobResponse(BaseModel):
    job: Job = Field(description="Complete job information.")


class NextPageLink(BaseModel):
    next: URL = Field(description="URL for getting the next page of results.")


class Jobs(BaseModel):
    jobs: list[Job] = Field(description="List of jobs.")
    links: Optional[NextPageLink] = Field(default=None, alias="_links",
                                          description="Contains the hyperlink to the next page of results, "
                                                      "if applicable.")

    def get_continuation_token(self) -> Optional[str]:
        """
        Return continuation token for next query

        :return: Continuation Token value if any
        """
        if not self.links:
            return None
        parsed = urllib.parse.urlparse(self.links.next.href)
        params = urllib.parse.parse_qs(parsed.query)
        cts = params.get("continuationToken")
        if cts is None:
            return None
        return cts[0]


class Progress(BaseProgress):
    pass


class ProgressResponse(BaseModel):
    progress: Progress = Field(description="Progress information.")


class Message(BaseModel):
    code: str = Field(description="Unique identifier for an error.")
    title: str = Field(description="Title of the error.")
    message: str = Field(description="Message of the error.")
    params: list[str] = Field(description="Parameters to be placed in the message. "
                                          "Can be used for localization effort.")


class Messages(BaseModel):
    errors: list[Message] = Field(description="List of potential errors from the job execution.")
    warnings: list[Message] = Field(description="List of potential warnings from the job execution.")


class MessagesResponse(BaseModel):
    messages: Messages = Field(description="Messages.")

File: typescript/packages/reality-capture/src/service/job.ts
import { z } from "zod";

import { CalibrationSpecificationsCreateSchema, CalibrationSpecificationsSchema } from "../specifications/calibration";
//import { ChangeDetectionSpecificationsCreateSchema, ChangeDetectionSpecificationsSchema } from '../specifications/change_detection';
import { ConstraintsSpecificationsCreateSchema, ConstraintsSpecificationsSchema } from "../specifications/constraints";
import { FillImagePropertiesSpecificationsCreateSchema, FillImagePropertiesSpecificationsSchema } from "../specifications/fill_image_properties";
import { ImportPCSpecificationsCreateSchema, ImportPCSpecificationsSchema } from "../specifications/import_point_cloud";
//import { Objects2DSpecificationsCreateSchema, Objects2DSpecificationsSchema } from '../specifications/objects2d';
import { ProductionSpecificationsCreateSchema, ProductionSpecificationsSchema } from "../specifications/production";
import { ReconstructionSpecificationsCreateSchema, ReconstructionSpecificationsSchema } from "../specifications/reconstruction";
//import { Segmentation2DSpecificationsCreateSchema, Segmentation2DSpecificationsSchema } from '../specifications/segmentation2d';
//import { Segmentation3DSpecificationsCreateSchema, Segmentation3DSpecificationsSchema } from '../specifications/segmentation3d';
//import { SegmentationOrthophotoSpecificationsCreateSchema, SegmentationOrthophotoSpecificationsSchema } from '../specifications/segmentation_orthophoto';
import { TilingSpecificationsCreateSchema, TilingSpecificationsSchema } from "../specifications/tiling";
import { TouchUpExportSpecificationsCreateSchema, TouchUpImportSpecificationsCreateSchema, TouchUpExportSpecificationsSchema, TouchUpImportSpecificationsSchema } from "../specifications/touchup";
import { WaterConstraintsSpecificationsCreateSchema, WaterConstraintsSpecificationsSchema } from "../specifications/water_constraints";
//import { TrainingO2DSpecificationsCreateSchema, TrainingS3DSpecificationsCreateSchema, TrainingO2DSpecificationsSchema, TrainingS3DSpecificationsSchema } from '../specifications/training';
//import { PointCloudConversionSpecificationsCreateSchema, PointCloudConversionSpecificationsSchema } from '../specifications/point_cloud_conversion';
import { GaussianSplatsSpecificationsCreateSchema, GaussianSplatsSpecificationsSchema } from "../specifications/gaussian_splats";
import { URLSchema } from "./bucket";
//import { EvalO2DSpecificationsCreateSchema, EvalO2DSpecificationsSchema } from '../specifications/eval_o2d';
//import { EvalO3DSpecificationsCreateSchema, EvalO3DSpecificationsSchema } from '../specifications/eval_o3d';
//import { EvalS2DSpecificationsCreateSchema, EvalS2DSpecificationsSchema } from '../specifications/eval_s2d';
//import { EvalS3DSpecificationsCreateSchema, EvalS3DSpecificationsSchema } from '../specifications/eval_s3d';
//import { EvalSOrthoSpecificationsCreateSchema, EvalSOrthoSpecificationsSchema  } from '../specifications/eval_sortho';


export enum JobType {
  CALIBRATION = "Calibration",
  //CHANGE_DETECTION = "ChangeDetection",
  CONSTRAINTS = "Constraints",
  FILL_IMAGE_PROPERTIES = "FillImageProperties",
  IMPORT_POINT_CLOUD = "ImportPointCloud",
  //OBJECTS_2D = "Objects2D",
  PRODUCTION = "Production",
  RECONSTRUCTION = "Reconstruction",
  //SEGMENTATION_2D = "Segmentation2D",
  //SEGMENTATION_3D = "Segmentation3D",
  //SEGMENTATION_ORTHOPHOTO = "SegmentationOrthophoto",
  TILING = "Tiling",
  TOUCH_UP_IMPORT = "TouchUpImport",
  TOUCH_UP_EXPORT = "TouchUpExport",
  WATER_CONSTRAINTS = "WaterConstraints",
  //TRAINING_O2D = "TrainingO2D",
  //TRAINING_S3D = "TrainingS3D",
  //EVAL_O2D = "EvalO2D",
  //EVAL_O3D = "EvalO3D",
  //EVAL_S2D = "EvalS2D",
  //EVAL_S3D = "EvalS3D",
  //EVAL_SORTHO = "EvalSOrtho",
  //POINT_CLOUD_CONVERSION = "PointCloudConversion",
  GAUSSIAN_SPLATS = "GaussianSplats"
}

export enum Service {
  MODELING = "Modeling",
  ANALYSIS = "Analysis",
  CONVERSION = "Conversion"
}

export function getAppropriateService(jt: JobType): Service {
  if ([
    JobType.FILL_IMAGE_PROPERTIES,
    JobType.IMPORT_POINT_CLOUD,
    JobType.CALIBRATION,
    JobType.TILING,
    JobType.PRODUCTION,
    JobType.RECONSTRUCTION,
    JobType.CONSTRAINTS,
    JobType.TOUCH_UP_EXPORT,
    JobType.TOUCH_UP_IMPORT,
    JobType.WATER_CONSTRAINTS,
    JobType.GAUSSIAN_SPLATS
  ].includes(jt)) {
    return Service.MODELING;
  }
  /*if ([
  JobType.OBJECTS_2D,
  JobType.SEGMENTATION_2D,
  JobType.SEGMENTATION_3D,
  JobType.SEGMENTATION_ORTHOPHOTO,
  JobType.CHANGE_DETECTION,
  JobType.TRAINING_O2D,
  JobType.EVAL_O2D,
  JobType.EVAL_O3D,
  JobType.EVAL_S2D,
  JobType.EVAL_S3D,
  JobType.EVAL_SORTHO
].includes(jt)) {
  return Service.ANALYSIS;
}
return Service.CONVERSION;*/
  throw new Error("Other job types are not implemented yet");
}

export enum JobState {
  QUEUED = "Queued",
  ACTIVE = "Active",
  SUCCESS = "Success",
  FAILED = "Failed",
  TERMINATING_ON_CANCEL = "TerminatingOnCancel",
  TERMINATING_ON_FAILURE = "TerminatingOnFailure",
  CANCELLED = "Cancelled"
}

export const JobCreateSchema = z.object({
  name: z.string().min(3).optional().describe("Displayable job name."),
  type: z.nativeEnum(JobType).describe("Type of job."),
  specifications: z.union([
    CalibrationSpecificationsCreateSchema,
    //ChangeDetectionSpecificationsCreateSchema,
    ConstraintsSpecificationsCreateSchema,
    FillImagePropertiesSpecificationsCreateSchema,
    ImportPCSpecificationsCreateSchema,
    //Objects2DSpecificationsCreateSchema,
    ProductionSpecificationsCreateSchema,
    ReconstructionSpecificationsCreateSchema,
    //Segmentation2DSpecificationsCreateSchema,
    //Segmentation3DSpecificationsCreateSchema,
    //SegmentationOrthophotoSpecificationsCreateSchema,
    TilingSpecificationsCreateSchema,
    TouchUpExportSpecificationsCreateSchema,
    TouchUpImportSpecificationsCreateSchema,
    WaterConstraintsSpecificationsCreateSchema,
    //TrainingO2DSpecificationsCreateSchema,
    //PointCloudConversionSpecificationsCreateSchema,
    //TrainingS3DSpecificationsCreateSchema,
    GaussianSplatsSpecificationsCreateSchema,
    //EvalO2DSpecificationsCreateSchema,
    //EvalO3DSpecificationsCreateSchema,
    //EvalS2DSpecificationsCreateSchema,
    //EvalS3DSpecificationsCreateSchema,
    //EvalSOrthoSpecificationsCreateSchema
  ]).describe("Specifications aligned with the job type."),
  iTwinId: z.string().describe("iTwin ID, used by the service for finding input reality data and uploading output data."),
});
export type JobCreate = z.infer<typeof JobCreateSchema>;

export const ExecutionSchema = z.object({
  createdDateTime: z.coerce.date().describe("Creation date time for the job."),
  startedDateTime: z.coerce.date().nullable().optional().describe("Start date time for the job."),
  endedDateTime: z.coerce.date().nullable().optional().describe("End date time for the job."),
  processingUnits: z.number().nullable().optional().describe("Processing units consumed by the job."),
});
export type Execution = z.infer<typeof ExecutionSchema>;

const CommonFields = {
  id: z.string().describe("Job unique identifier."),
  name: z.string().min(3).optional().describe("Displayable job name."),
  itwinId: z.string().describe("iTwin ID, used by the service for finding input reality data and uploading output data."),
  state: z.nativeEnum(JobState).describe("State of the job."),
  executionInfo: ExecutionSchema.describe("Known execution information for the job."),
  userId: z.string().describe("Identifier of the user that created the job."),
};

export const JobSchema = z.discriminatedUnion("type", [
  z.object({
    ...CommonFields,
    type: z.literal("Calibration"),
    specifications: CalibrationSpecificationsSchema,
  }),
  /*z.object({
  ...CommonFields,
  type: z.literal("ChangeDetection"),
  specifications: ChangeDetectionSpecificationsSchema,
}),*/
  z.object({
    ...CommonFields,
    type: z.literal("Constraints"),
    specifications: ConstraintsSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("FillImageProperties"),
    specifications: FillImagePropertiesSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("ImportPointCloud"),
    specifications: ImportPCSpecificationsSchema,
  }),
  /*z.object({
  ...CommonFields,
  type: z.literal("Objects2D"),
  specifications: Objects2DSpecificationsSchema,
}),*/
  z.object({
    ...CommonFields,
    type: z.literal("Production"),
    specifications: ProductionSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("Reconstruction"),
    specifications: ReconstructionSpecificationsSchema,
  }),
  /*z.object({
  ...CommonFields,
  type: z.literal("Segmentation2D"),
  specifications: Segmentation2DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("Segmentation3D"),
  specifications: Segmentation3DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("SegmentationOrthophoto"),
  specifications: SegmentationOrthophotoSpecificationsSchema,
}),*/
  z.object({
    ...CommonFields,
    type: z.literal("Tiling"),
    specifications: TilingSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("TouchUpExport"),
    specifications: TouchUpExportSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("TouchUpImport"),
    specifications: TouchUpImportSpecificationsSchema,
  }),
  z.object({
    ...CommonFields,
    type: z.literal("WaterConstraints"),
    specifications: WaterConstraintsSpecificationsSchema,
  }),
  /*z.object({
  ...CommonFields,
  type: z.literal("TrainingO2D"),
  specifications: TrainingO2DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("TrainingS3D"),
  specifications: TrainingS3DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("PointCloudConversion"),
  specifications: PointCloudConversionSpecificationsSchema,
}),*/
  z.object({
    ...CommonFields,
    type: z.literal("GaussianSplats"),
    specifications: GaussianSplatsSpecificationsSchema,
  }),
  /*z.object({
  ...CommonFields,
  type: z.literal("EvalO2D"),
  specifications: EvalO2DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("EvalO3D"),
  specifications: EvalO3DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("EvalS2D"),
  specifications: EvalS2DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("EvalS3D"),
  specifications: EvalS3DSpecificationsSchema,
}),
z.object({
  ...CommonFields,
  type: z.literal("EvalSOrtho"),
  specifications: EvalSOrthoSpecificationsSchema,
}),*/
]);
export type Job = z.infer<typeof JobSchema>;

export const NextLinkSchema = z.object({
  next: URLSchema.describe("URL for getting the next page of results.")
});
export type NextLink = z.infer<typeof NextLinkSchema>;

export const JobsSchema = z.object({
  jobs: z.array(JobSchema).describe("List of jobs."),
  _links: NextLinkSchema.describe("Contains the hyperlink to the next page of results, if applicable."),
});
export type Jobs = z.infer<typeof JobsSchema>;

export const JobResponseSchema = z.object({
  job: JobSchema.describe("Complete job information."),
});
export type JobResponse = z.infer<typeof JobResponseSchema>;

export const ProgressSchema = z.object({
  state: z.nativeEnum(JobState).describe("State of the job."),
  percentage: z.number().min(0).max(100).describe("Progress of the job."),
});
export type Progress = z.infer<typeof ProgressSchema>;

export const ProgressResponseSchema = z.object({
  progress: ProgressSchema.describe("Progress information."),
});
export type ProgressResponse = z.infer<typeof ProgressResponseSchema>;

export const MessageSchema = z.object({
  code: z.string().describe("Unique identifier for an error."),
  title: z.string().describe("Title of the error."),
  message: z.string().describe("Message of the error."),
  params: z.array(z.string()).describe("Parameters to be placed in the message. Can be used for localization effort."),
});
export type Message = z.infer<typeof MessageSchema>;

export const MessagesSchema = z.object({
  errors: z.array(MessageSchema).describe("List of potential errors from the job execution."),
  warnings: z.array(MessageSchema).describe("List of potential warnings from the job execution."),
});
export type Messages = z.infer<typeof MessagesSchema>;

export const MessagesResponseSchema = z.object({
  messages: MessagesSchema.describe("Messages."),
});
export type MessagesResponse = z.infer<typeof MessagesResponseSchema>;